crates/durability/src/imp/local.rs

```diff
@@ -0,0 +1,277 @@
+use std::{
+    io,
+    num::NonZeroU16,
+    panic,
+    path::PathBuf,
+    sync::{
+        atomic::{
+            AtomicI64, AtomicU64,
+            Ordering::{Relaxed, SeqCst},
+        },
+        Arc,
+    },
+    time::Duration,
+};
+
+use anyhow::Context as _;
+use log::{debug, info, trace, warn};
+use spacetimedb_commitlog::{error, payload::Txdata, Commit, Commitlog, Decoder, Encode, Transaction};
+use tokio::{
+    sync::mpsc,
+    task::{spawn_blocking, JoinHandle},
+};
+use tracing::instrument;
+
+use crate::{Durability, History, TxOffset};
+
+/// [`Local`] configuration.
+#[derive(Clone, Copy, Debug)]
+pub struct Options {
+    /// Periodically flush and sync the log this often.
+    ///
+    /// Default: 500ms
+    pub sync_interval: Duration,
+    /// [`Commitlog`] configuration.
+    pub commitlog: spacetimedb_commitlog::Options,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            sync_interval: Duration::from_millis(500),
+            commitlog: Default::default(),
+        }
+    }
+}
+
+/// [`Durability`] implementation backed by a [`Commitlog`] on local storage.
+///
+/// The commitlog is constrained to store the canonical [`Txdata`] payload,
+/// where the generic parameter `T` is the type of the row data stored in
+/// the mutations section.
+///
+/// `T` is left generic in order to allow bypassing the `ProductValue`
+/// intermediate representation in the future.
+///
+/// Note, however, that instantiating `T` to a different type may require to
+/// change the log format version!
+pub struct Local<T> {
+    clog: Arc<Commitlog<Txdata<T>>>,
+    offset: Arc<AtomicI64>,
+    queue: mpsc::UnboundedSender<Txdata<T>>,
+    queue_depth: Arc<AtomicU64>,
```

Comment:
Could you add comments to the atomic members here describing what (if anything) they're used to synchronize, and what `Ordering` accomplishes that? It feels intuitive to me that `queue_depth` is just a shared counter for diagnostic/reporting purposes, and therefore can be `Relaxed`, but it's not obvious to me why the `offset` needs to be `SeqCst`.

Reply:
The offset would be used by the datastore to squash transactions which became durable, and don’t need to be hidden from readers which requested durable-only data.

It seemed reasonable to keep this strongly ordered so as to not hang on to data longer than necessary.

But, I can be convinced of this being flawed reasoning. Wdyt?

Reply:
Updated to acquire / release and added commentary.